### PR TITLE
AndroidIA: add more formats in i915_private

### PR DESCRIPTION
**Pull request by js0701**
Those formats with special layout are potencially
need by camera usage and encode can take them as
input.
JIRA: None
Test: None

Signed-off-by: Johnson Lin <johnson.lin@intel.com>

### CODE REVIEW

**Comment by tpalli** [comment]
Uh huh ... these are non-standard enums and it will be very hard to remove them later. I would really encourage you guys to tell multimedia team to fix their driver. Also fact that linear vs tiled is present in these enums is simply horrible, these should be implementation details handled by other API such as the new 'modifer' API for dmabufs.

File: i915_private.c [hunk 1]
```diff
@@ -90,15 +99,29 @@ int i915_private_add_combinations(struct driver *drv)
...
+	  return ret;
 
 	metadata.tiling = I915_TILING_Y;
 	metadata.priority = 3;
 	metadata.modifier = I915_FORMAT_MOD_Y_TILED;
-	drv_add_combinations(drv, private_source_formats, ARRAY_SIZE(private_source_formats),
-			     &metadata, texture_flags | BO_USE_CAMERA_MASK);
+	ret = drv_add_combinations(drv, intel_private_ytiled_source_formats,
+		                 ARRAY_SIZE(intel_private_ytiled_source_formats), &metadata,
+	                     BO_USE_CAMERA_MASK);
```
Comment (kalyankondapally) [inline i915_private.c]: You should be able to texture from it, I think we need the texture_flags here

File: i915_private.c [hunk 2]
```diff
@@ -147,6 +176,151 @@ uint32_t i915_private_bpp_from_format(uint32_t format, size_t plane)
 	return 0;
 }
 
+bool  i915_private_is_private_format(uint32_t format)
+{
+  //As DRM_FORMAT_NV12_Y_TILED_INTEL is already caculated so excluded here
+  if(format > DRM_FORMAT_NV12_Y_TILED_INTEL && format <= DRM_FORMAT_BGRP_INTEL)
```
Comment (kalyankondapally) [inline i915_private.c]: We probably want to unify DRM_FORMAT_NV12_Y_TILED_INTEL also, so all are in one place

**Comment by kalyankondapally** [comment]
@tpalli Ya, for now we have to support to them to get things enabled. Than we start cleaning up things which no body is dependent upon.

File: i915_private.c [hunk 3]
```diff
@@ -147,6 +176,151 @@ uint32_t i915_private_bpp_from_format(uint32_t format, size_t plane)
...
+	 case DRM_FORMAT_YVU422_H_INTEL:
+	 case DRM_FORMAT_YUV422_H_INTEL:
+	 case DRM__FORMAT_GENERIC_8BIT_INTEL:
+	 case DRM_FORMAT_YUV411_INTEL:
+	 case DRM_FORMAT_YUV444_INTEL:
+	 case DRM_FORMAT_RGBP_INTEL:
+	 case DRM_FORMAT_BGRP_INTEL:
+	   aligned_height = ALIGN(height, 64);
+	   break;
+
```
Comment (kalyankondapally) [inline i915_private.c]: extra line ?

File: i915_private.c (hunk 1 above)
Comment (js0701) [reply i915_private.c]: OK

File: i915_private.c (hunk 2 above)
Comment (js0701) [reply i915_private.c]: OK, will do it. If so, there will be some code clean up

File: i915_private.c (hunk 3 above)
Comment (js0701) [reply i915_private.c]: Will remove

**Comment by js0701** [comment]
Not just the Xtile and Ytile problem. In those formats, the alignement and offset are different from what is standard. If we use modificators, we probably need to add modifactor beyond standard one
@tpalli

**Comment by tpalli** [comment]
that should be ok, modifiers are designed so that there can be more and these can be vendor specific

**Comment by js0701** [comment]
Updated PR @kalyankondapally

File: cros_gralloc/i915_private_android.cc [hunk 4]
```diff
@@ -80,18 +106,28 @@ int32_t i915_private_invert_format(int format)
...
-	case HAL_PIXEL_FORMAT_NV12_Y_TILED_INTEL:
 	case HAL_PIXEL_FORMAT_YCbCr_422_I:
 	case HAL_PIXEL_FORMAT_YCbCr_422_888:
 	case HAL_PIXEL_FORMAT_YCbCr_444_888:
 	case HAL_PIXEL_FORMAT_YCrCb_420_SP:
 	case HAL_PIXEL_FORMAT_Y16:
 		return true;
 	default:
+	    if(i915_private_is_private_yuv_format(droid_format))
+		    return true;
```
Comment (jekstrand) [inline cros_gralloc/i915_private_android.cc]: Indentation is messed up here.  This file appears to use tabs but you've got mixed tabs and spaces in here.

File: drv.c [hunk 5]
```diff
@@ -12,11 +12,10 @@
 #include <string.h>
 #include <sys/mman.h>
 #include <xf86drm.h>
-
+#include "i915_private.h"
 #include "drv_priv.h"
 #include "helpers.h"
 #include "util.h"
-#include "i915_private.h"
```
Comment (jekstrand) [inline drv.c]: Why are includes moving around?

File: drv.c [hunk 6]
```diff
@@ -508,6 +507,7 @@ uint32_t drv_resolve_format(struct driver *drv, uint32_t format, uint64_t usage)
 
 size_t drv_num_planes_from_format(uint32_t format)
 {
+
```
Comment (jekstrand) [inline drv.c]: Don't need the extra line

File: helpers.c [hunk 7]
```diff
@@ -13,7 +13,6 @@
 #include <sys/mman.h>
 #include <xf86drm.h>
 #include <xf86drmMode.h>
-
```
Comment (jekstrand) [inline helpers.c]: Spurious change

File: helpers.c [hunk 8]
```diff
@@ -100,7 +105,6 @@ static uint32_t bpp_from_format(uint32_t format, size_t plane)
 	case DRM_FORMAT_XRGB8888:
 		return 32;
 	}
-
```
Comment (jekstrand) [inline helpers.c]: Spurious change

File: i915.c [hunk 9]
```diff
@@ -322,21 +320,24 @@ static int i915_bo_create(struct bo *bo, uint32_t width, uint32_t height, uint32
 
 	bo->tiling = combo->metadata.tiling;
 
-	stride = drv_stride_from_format(format, width, 0);
+	if(i915_private_is_private_format(format)) {
+	  i915_private_drv_bo_from_format(bo, width, height, format);
+	} else {
+	  stride = drv_stride_from_format(format, width, 0);
```
Comment (jekstrand) [inline i915.c]: Again, whitespace here doesn't make any sense.  Should be tabs.

File: i915.c [hunk 10]
```diff
@@ -322,21 +320,24 @@ static int i915_bo_create(struct bo *bo, uint32_t width, uint32_t height, uint32
...
-        }
+    if (flags & BO_USE_CURSOR) {
+      width = ALIGN(width, i915_dev->cursor_width);
+      height = ALIGN(height, i915_dev->cursor_height);
+      stride = drv_stride_from_format(format, width, 0);
+    } else {
+      ret = i915_align_dimensions(bo, bo->tiling, &stride, &height);
+      if (ret)
+        return ret;
+    }
```
Comment (jekstrand) [inline i915.c]: This change seems to have nothing to do with what the patch is trying to accomplish.  it just takes the whitespace from agruably wrong (it's using spaces instead of tabs and has some weird indentation) to different but probably still wrong.  Since you don't actually have to change this code, maybe it's best left alone.

File: i915.c [hunk 11]
```diff
@@ -354,16 +355,17 @@ static int i915_bo_create(struct bo *bo, uint32_t width, uint32_t height, uint32
...
-
-		drv_bo_from_format(bo, stride, height, DRM_FORMAT_YVU420);
-		total_size = bo->total_size;
-		drv_bo_from_format(bo, stride, unaligned_height, format);
-		bo->total_size = total_size;
-	} else {
-		drv_bo_from_format(bo, stride, height, format);
+		if (format == DRM_FORMAT_YVU420_ANDROID) {
+			uint32_t unaligned_height = bo->height;
+	    size_t total_size;
```
Comment (jekstrand) [inline i915.c]: Mixed tabs and spaces again.

File: i915.c [hunk 12]
```diff
@@ -379,8 +381,8 @@ static int i915_bo_create(struct bo *bo, uint32_t width, uint32_t height, uint32
         /*
          * Ensure we pass aligned width/height.
          */
-        bo->width = width;
-        bo->height = height;
+	bo->width = width;
+	bo->height = height;
```
Comment (jekstrand) [inline i915.c]: This is a good change in that it makes the whitespace correct here but I don't think it belongs in this patch.  Maybe break it out into it's own.

File: i915_private.c [hunk 13]
```diff
@@ -90,15 +99,29 @@ int i915_private_add_combinations(struct driver *drv)
 	metadata.tiling = I915_TILING_NONE;
 	metadata.priority = 1;
 	metadata.modifier = DRM_FORMAT_MOD_NONE;
-	drv_add_combinations(drv, private_linear_source_formats,
+	ret = drv_add_combinations(drv, private_linear_source_formats,
 			     ARRAY_SIZE(private_linear_source_formats), &metadata,
 			     texture_flags | BO_USE_CAMERA_MASK);
+	if (ret)
+	  return ret;
```
Comment (jekstrand) [inline i915_private.c]: Tabs vs. spaces again.

File: i915_private.c [hunk 14]
```diff
@@ -90,15 +99,29 @@ int i915_private_add_combinations(struct driver *drv)
...
+
+	ret = drv_add_combinations(drv, intel_private_linear_source_formats,
+		         ARRAY_SIZE(intel_private_linear_source_formats), &metadata,
+		         texture_flags | BO_USE_CAMERA_MASK);
+	if (ret)
+	  return ret;
```
Comment (jekstrand) [inline i915_private.c]: Tabs vs. spaces

File: i915_private.c [hunk 15]
```diff
@@ -90,15 +99,29 @@ int i915_private_add_combinations(struct driver *drv)
...
 	metadata.tiling = I915_TILING_Y;
 	metadata.priority = 3;
 	metadata.modifier = I915_FORMAT_MOD_Y_TILED;
-	drv_add_combinations(drv, private_source_formats, ARRAY_SIZE(private_source_formats),
-			     &metadata, texture_flags | BO_USE_CAMERA_MASK);
+	ret = drv_add_combinations(drv, intel_private_ytiled_source_formats,
+		                 ARRAY_SIZE(intel_private_ytiled_source_formats), &metadata,
+	                   texture_flags | BO_USE_CAMERA_MASK);
+	if (ret)
+	  return ret;
```
Comment (jekstrand) [inline i915_private.c]: Tabs vs. spaces

File: i915_private.c [hunk 16]
```diff
@@ -90,15 +99,29 @@ int i915_private_add_combinations(struct driver *drv)
...
 
 	metadata.tiling = I915_TILING_Y;
 	metadata.priority = 3;
 	metadata.modifier = I915_FORMAT_MOD_Y_TILED;
-	drv_add_combinations(drv, private_source_formats, ARRAY_SIZE(private_source_formats),
-			     &metadata, texture_flags | BO_USE_CAMERA_MASK);
+	ret = drv_add_combinations(drv, intel_private_ytiled_source_formats,
+		                 ARRAY_SIZE(intel_private_ytiled_source_formats), &metadata,
+	                   texture_flags | BO_USE_CAMERA_MASK);
```
Comment (jekstrand) [inline i915_private.c]: The indentation here doesn't look quite right.

File: i915_private.c [hunk 17]
```diff
@@ -147,6 +167,152 @@ uint32_t i915_private_bpp_from_format(uint32_t format, size_t plane)
...
+  if(format >= DRM_FORMAT_NV12_Y_TILED_INTEL && format <= DRM_FORMAT_BGRP_INTEL)
+	  return true;
+  else
+	  return false;
+}
+
+
+// This function converts from private formats to standard drm formats
+// If we want mesa to support those formats we need to convert it to standard drm formats
+void i95_private_override_format(struct bo *bo)
```
Comment (jekstrand) [inline i915_private.c]: I think you want this to be i915 not i95

File: i915_private.c [hunk 18]
```diff
@@ -147,6 +167,152 @@ uint32_t i915_private_bpp_from_format(uint32_t format, size_t plane)
...
+	      else if (width <= 1280)  stride = 1280;
+	      else if (width <= 2048)  stride = 2048;
+	      else if (width <= 4096)  stride = 4096;
+      break;
+	  case DRM_FORMAT_NV12_LINEAR_PACKED_INTEL:
+	      stride = ALIGN(stride, 2);
+	      break;
+	default:
+		break;
+	}
```
Comment (jekstrand) [inline i915_private.c]: This isn't indented right either.

File: i915_private.c [hunk 19]
```diff
@@ -147,6 +167,152 @@ uint32_t i915_private_bpp_from_format(uint32_t format, size_t plane)
...
+	 case DRM_FORMAT_RGBP_INTEL:
+	 case DRM_FORMAT_BGRP_INTEL:
+	     aligned_height = ALIGN(height, 64);
+	     break;
+	 case DRM_FORMAT_NV12_LINEAR_PACKED_INTEL:
+	     aligned_height = ALIGN(height, 2);
+	     if(plane >= 1)
+	         aligned_height /= 2;
+		 break;
+	 default:
```
Comment (jekstrand) [inline i915_private.c]: Neither is most of this switch.

File: i915_private.c [hunk 20]
```diff
@@ -185,8 +363,8 @@ uint32_t i915_private_resolve_format(uint32_t format, uint64_t usage, uint32_t *
 		/* KBL camera subsystem requires NV12. */
 		if (usage & (BO_USE_CAMERA_READ | BO_USE_CAMERA_WRITE)) {
 			*resolved_format = DRM_FORMAT_NV12;
-                        return 1;
-                }
+			return 1;
+		}
```
Comment (jekstrand) [inline i915_private.c]: This looks like a good change in that it fixes whitespace to be tabs.  However, it doesn't really belong in this patch.  It may be a good idea to have a separate "fix whitespace" patch to fix these issues.

File: cros_gralloc/i915_private_android.cc [hunk 21]
```diff
@@ -13,6 +13,7 @@
 
 #include <hardware/gralloc.h>
 
+
```
Comment (jekstrand) [inline cros_gralloc/i915_private_android.cc]: This extra new line isn't needed.

**Comment by jekstrand** [comment]
@js0701 wrote:
> Not just the Xtile and Ytile problem. In those formats, the alignement and offset are different from what is standard. If we use modificators, we probably need to add modifactor beyond standard one
@tpalli

I don't think that's true.  Yes, using Y-tiling implies a certain alignment on buffer creation, but that information can just as easily be carried by DRM_FORMAT_MODIFIER_Y_TILED.  It already implies that each of the planes needs to be tile-aligned.

@kalyankondapally wrote:
> @tpalli Ya, for now we have to support them get things enabled. Than we start cleaning up things which no body is dependent upon. We are tyring to obsorb everything here so we dont spread these across components, only thing which would need mapping in other components would be NV12_Y_TILED HAL format.

I share the concern with @tpalli here.  Yes, we may need to cram some stuff in now to get camera working on AndroidIA.  I'm not sure how things work in the intel minigbm branch, but in mesa and other parts of the linux ecosystem, these things are permanent ABI and we can't just "clean up later".  Once you start exposing them, there's no stopping.  For that reason, I really don't want to support these Intel-specific formats in upstream mesa when we can just as easily do the same thing with DRM_FORMAT_NV12 + I915_FORMAT_MOD_Y_TILED.

File: i915.c (hunk 9 above)
Comment (kalyankondapally) [reply i915.c]: @js0701 "Ensure that your change conforms with the Google C++ style guide. Running the following command before commit will show you if any changes are needed:

git diff | clang-format-diff-3.5 -p 1 -style=file

If you are happy with the changes, running following command will apply the changes:

git diff | clang-format-diff-3.5 -p 1 -style=file -i"

Can you run these and clean it up.

**Comment by kalyankondapally** [comment]
@jekstrand  Problem is Android doesn't have a way to pass the modifiers when allocating buffers. These formats dont need support in Mesa side. Most of them are for optimizing case for Media/Camera while producing content and will eventually go through an intermediate conversion pass in libva side. Still checking which of them are really needed to be enabled here.

File: helpers.c (hunk 8 above)
Comment (js0701) [reply helpers.c]: @kalyankondapally  code clean done

File: i915_private.c (hunk 20 above)
Comment (js0701) [reply i915_private.c]: Thanks @jekstrand for the comments
Anyway I am following Kalyan's command to clean code.

File: helpers.c (hunk 8 above)
Comment (js0701) [reply helpers.c]: bpp_from_format for plane >0 is never used in mingbm actually.  Restore the change since it is not relevant here

**Comment by js0701** [comment]
@kalyankondapally 
Maybe we can put these things in a new branch.
When we meet new requirements from Android then we pick it.
This can avoid clean unused things

**Comment by kalyankondapally** [comment]
@js0701 If these are not needed, we should just drop them off and not encourage the usage as concerns raised here. (i.e. Dont want to have them in a branch either.)Only one case is failing now(I will share the test case offline), if it doesn't need these changes than we can just drop this PR.

**Comment by jekstrand** [comment]
@kalyankondapally Right.  Somehow gralloc needs to turn a HAL_PIXEL_FORMAT and set of usage bits into a DRM format and modifier.  (I don't know how our gralloc is structured so I can't really comment on how at the moment.)  Once we've determined a modifier and allocated the memory, we should be able to attach the modifier to the ANativeBuffer using a couple of those opaque integers.  Am I missing something here?  My knowledge of Android internals is very rusty and wasn't all that great at the time.

As far as who is using these extra formats, yes, I understand that they're used by camera and media.  However, I also know that there are people working on NV12 scanout and, if you want to pass a Y-tiled NV12 surface to the kernel for scanout, it's going to be done using a modifier.  Also, if we want 3D to be efficient, we need to get CCS scanout plumbed through surface flinger and IAHardwareComposer and both the kernel and mesa interfaces for dealing with CCS surface are based on modifiers.

The point I'm trying to make (I'm sorry if I'm making it a bit strongly here) is that AndroidIA will have to start supporting modifiers eventually.  It may not be today and it may not be next month, but we need to start moving that direction.  As I said above, adding a pile of formats now may be a decent pragmatic solution but, in the long term, I think we want to convert the camera and media blocks to use modifiers for specifying things like tiling layouts.

**Comment by kalyankondapally** [comment]
@jekstrand Agree with you, all valid points. The problem we where trying to solve here was some non standard "internal" HAL formats, not trying to avoid the usage of Modifiers. Passing the needed modifier information to IAHardwareComposer via SF should be pretty straightforward, as part of Gralloc attributes here https://github.com/intel/minigbm/blob/master/cros_gralloc/cros_gralloc_handle.h#L26 ? Looks like https://www.khronos.org/registry/EGL/extensions/EXT/EGL_EXT_image_dma_buf_import_modifiers.txt 
is not yet supported for Android on Mesa. That's another piece which needs to be resolved.

We managed to satisfy the needs of Media, Camera without these changes, so we should be fine now and don't need this PR.

**Comment by js0701** [comment]
Obsolete
